### PR TITLE
[CDAP-18648] Terminate the program on client side

### PR DESCRIPTION
**Pull request by snehalakshmisha**
Stop the program on client side when we receive stopping status from RuntimeHandler
Jira: https://cdap.atlassian.net/browse/CDAP-18744

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/runtimejob/DefaultRuntimeJob.java [hunk 1]
```diff
@@ -267,6 +268,11 @@ public void run(RuntimeJobEnvironment runtimeJobEnv) throws Exception {
       try (Program program = createProgram(cConf, programRunner, programDescriptor, programOpts)) {
         ProgramController controller = programRunner.run(program, programOpts);
         controllerFuture.complete(controller);
+        runtimeClientService.acceptStopProgramCommandAsync(isGracefulStopRequested -> {
+          if (isGracefulStopRequested) {
+            controller.stop();
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/runtimejob/DefaultRuntimeJob.java]: so if it is not a graceful stop, then what is it? And why don't stop if it is not graceful??

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClient.java [hunk 2]
```diff
@@ -122,14 +127,29 @@ public void sendMessages(ProgramRunId programRunId,
       try (Reader reader = new InputStreamReader(urlConn.getInputStream(), StandardCharsets.UTF_8)) {
         ProgramRunInfo responseBody = GSON.fromJson(reader, ProgramRunInfo.class);
         if (responseBody.getProgramRunStatus() == ProgramRunStatus.STOPPING) {
-          // TODO - CDAP-18744 - terminate the program based on the program endtimestamp payload in response body
+          // this will trigger controller.stop() in DefaultRuntimeJob
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClient.java]: Remove this comment. What get triggered after the future completed is beyond this class knowledge.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClient.java [hunk 3]
```diff
@@ -63,6 +66,7 @@
 
   private final boolean compression;
   private final RemoteClient remoteClient;
+  private CompletableFuture<Boolean> future;
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClient.java]: Make fields `final` when possible.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClient.java [hunk 4]
```diff
@@ -122,14 +127,29 @@ public void sendMessages(ProgramRunId programRunId,
...
       closeURLConnection(urlConn);
     }
   }
 
+  /**
+   * Accepts a consumer that runs in a daemon thread
+   * The consumer is intended to be programController.stop() which stops the program on the client side
+   * @param stopper Consumer that runs in a daemon thread
+   */
+  public void acceptStopProgramCommandAsync(Consumer<Boolean> stopper) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClient.java]: Better call it `onProgramStopRequested`.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClient.java [hunk 5]
```diff
@@ -122,14 +127,29 @@ public void sendMessages(ProgramRunId programRunId,
...
         }
       }
     } finally {
       closeURLConnection(urlConn);
     }
   }
 
+  /**
+   * Accepts a consumer that runs in a daemon thread
+   * The consumer is intended to be programController.stop() which stops the program on the client side
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClient.java]: Doesn't need to say what the consumer is intended to do.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java [hunk 6]
```diff
@@ -144,6 +145,10 @@ long getProgramFinishTime() {
     return programFinishTime;
   }
 
+  public void acceptStopProgramCommandAsync(Consumer<Boolean> stopper) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java]: same as above

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java [hunk 7]
```diff
@@ -144,6 +144,10 @@ long getProgramFinishTime() {
     return programFinishTime;
   }
 
+  public void onProgramStopRequested(Runnable stopper) {
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java:151]: should we keep this package private? If not it should add javadoc for public method.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClient.java [hunk 8]
```diff
@@ -122,14 +125,27 @@ public void sendMessages(ProgramRunId programRunId,
...
         }
       }
     } finally {
       closeURLConnection(urlConn);
     }
   }
 
+  /**
+   * Accepts a consumer that runs in a daemon thread
+   * @param stopper Consumer that runs in a daemon thread
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClient.java]: Runnable that runs in daemon thread

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClient.java [hunk 9]
```diff
@@ -81,6 +83,7 @@ public RuntimeClient(CConfiguration cConf, RemoteClientFactory remoteClientFacto
     if (schema.getType() != Schema.Type.ARRAY || schema.getElementType().getType() != Schema.Type.BYTES) {
       throw new IllegalStateException("MonitorRequest schema should be an array of bytes");
     }
+    this.future = new CompletableFuture<>();
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClient.java]: Lets call it stopFuture or something similar for readability

**Review by CuriousVini — APPROVED** [review]
few minor comments. 

LGTM

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java (hunk 7 above)
Comment (snehalakshmisha) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java:151]: The DefaultRuntimeJob is outside of this package. Added a doc

**Pull request merged**